File: backend/app/schemas/trade.py
from datetime import date, datetime, time

from pydantic import BaseModel, ConfigDict, Field

from app.models.trade import TradeDirection
from app.models.trade_fill import TradeFillSide


class TradeFillWrite(BaseModel):
    side: TradeFillSide
    quantity: int = Field(gt=0)
    price: float = Field(gt=0)
    filled_at: datetime | None = None


class TradeBase(BaseModel):
    date: date
    ticker: str = Field(min_length=1)
    direction: TradeDirection
    entry_price: float
    exit_price: float
    quantity: int = Field(default=1, gt=0)
    setup_id: int | None = Field(default=None, gt=0)
    emotion_id: int | None = Field(default=None, gt=0)
    rule_followed: bool | None
    notes: str | None = None
    entry_time: time | None = None
    exit_time: time | None = None


class TradeCreate(TradeBase):
    rule_followed: bool
    pnl: float | None = None
    duration_seconds: int | None = Field(default=None, ge=0)
    use_fills: bool = False
    fills: list[TradeFillWrite] | None = None


class TradeBulkUpdateRequest(BaseModel):
    trade_ids: list[int] = Field(min_length=1)
    setup_id: int | None = Field(default=None, gt=0)
    emotion_id: int | None = Field(default=None, gt=0)
    rule_followed: bool | None = None


class TradeBulkUpdateResponse(BaseModel):
    updated_count: int
    errors: list[str] = Field(default_factory=list)


class TradePatch(BaseModel):
    setup_id: int | None = Field(default=None, gt=0)
    emotion_id: int | None = Field(default=None, gt=0)
    rule_followed: bool | None = None


class TradeFillRead(BaseModel):
    id: int
    trade_id: int
    filled_at: datetime | None
    side: TradeFillSide
    quantity: int
    price: float
    source: str | None
    source_id: str | None

    model_config = ConfigDict(from_attributes=True)


class TradeRead(TradeBase):
    id: int
    pnl: float
    contract_multiplier: int
    total_pnl_usd: float
    duration_seconds: int | None
    total_entry_qty: int | None = None
    total_exit_qty: int | None = None
    avg_entry_price: float | None = None
    avg_exit_price: float | None = None
    realized_pnl_usd: float | None = None
    is_partial: bool = False
    use_fills: bool = False
    fills: list[TradeFillRead] | None = None
    setup_name: str
    emotion_name: str
    source: str | None = None
    source_id: str | None = None

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/trade.ts
export type TradeDirection = "CALL" | "PUT";
export type TradeRuleFollowedFilter = "true" | "false" | "unknown";
export type TradeOutcomeFilter = "win" | "loss" | "breakeven";
export type TradeSourceFilter = "tos_csv" | "manual";
export type TradeClassificationFilter = "all" | "unclassified" | "classified";
export type TradeIndexBucketFilter = "first_3" | "after_3";
export type TradePatternFilter = "after_2_losses_next_trade";
export type TradeFillSide = "BUY" | "SELL";

export interface TradeListFilters {
  date?: string;
  start?: string;
  end?: string;
  import_batch_id?: number;
  ticker?: string;
  setup_id?: number;
  emotion_id?: number;
  rule_followed?: TradeRuleFollowedFilter;
  outcome?: TradeOutcomeFilter;
  source?: TradeSourceFilter;
  classification?: TradeClassificationFilter;
  entry_time_start_minute?: number;
  entry_time_end_minute?: number;
  hold_time_min_seconds?: number;
  hold_time_max_seconds?: number;
  trade_index_bucket?: TradeIndexBucketFilter;
  pattern?: TradePatternFilter;
  include_fills?: boolean;
}

export interface CreateTradeRequest {
  date: string;
  ticker: string;
  direction: TradeDirection;
  entry_price: number;
  exit_price: number;
  quantity: number;
  pnl?: number;
  setup_id?: number | null;
  emotion_id?: number | null;
  rule_followed: boolean;
  notes?: string | null;
  use_fills?: boolean;
  fills?: TradeFillWriteRequest[];
}

export interface BulkUpdateTradesRequest {
  trade_ids: number[];
  setup_id?: number | null;
  emotion_id?: number | null;
  rule_followed?: boolean | null;
}

export interface BulkUpdateTradesResponse {
  updated_count: number;
  errors: string[];
}

export interface TradePatchRequest {
  setup_id?: number | null;
  emotion_id?: number | null;
  rule_followed?: boolean | null;
}

export interface TradeResponse {
  id: number;
  date: string;
  ticker: string;
  direction: TradeDirection;
  entry_price: number;
  exit_price: number;
  pnl: number;
  quantity: number;
  contract_multiplier: number;
  total_pnl_usd: number;
  setup_id: number | null;
  setup_name: string;
  emotion_id: number | null;
  emotion_name: string;
  rule_followed: boolean | null;
  notes: string | null;
  source?: string | null;
  source_id?: string | null;
  entry_time?: string | null;
  exit_time?: string | null;
  duration_seconds?: number | null;
  total_entry_qty?: number | null;
  total_exit_qty?: number | null;
  avg_entry_price?: number | null;
  avg_exit_price?: number | null;
  realized_pnl_usd?: number | null;
  is_partial?: boolean;
  use_fills?: boolean;
  fills?: TradeFillResponse[] | null;
}

export interface TradeFillResponse {
  id: number;
  trade_id: number;
  filled_at: string | null;
  side: TradeFillSide;
  quantity: number;
  price: number;
  source: string | null;
  source_id: string | null;
}

export interface TradeFillWriteRequest {
  side: TradeFillSide;
  quantity: number;
  price: number;
  filled_at?: string | null;
}
